sound/soc/sof/intel/hda.c

```diff
@@ -607,18 +607,19 @@ int hda_dsp_probe(struct snd_sof_dev *sdev)
 	 * TODO: support interrupt mode selection with kernel parameter
 	 *       support msi multiple vectors
 	 */
-//	ret = pci_alloc_irq_vectors(pci, 1, 1, PCI_IRQ_MSI);
+	ret = pci_alloc_irq_vectors(pci, 1, 1, PCI_IRQ_MSI);
 	/* todo: MSI mode doesn't work for HDMI yet, debug it later */
-	ret = pci_alloc_irq_vectors(pci, 1, 1, PCI_IRQ_LEGACY);
 	if (ret < 0) {
 		dev_info(sdev->dev, "use legacy interrupt mode\n");
 		sdev->hda->irq = pci->irq;
 		sdev->ipc_irq = pci->irq;
+		sdev->msi_enabled = 0;
 	} else {
 		dev_info(sdev->dev, "use msi interrupt mode\n");
 		sdev->hda->irq = pci_irq_vector(pci, 0);
 		/* ipc irq number is the same of hda irq */
 		sdev->ipc_irq = sdev->hda->irq;
+		sdev->msi_enabled = 1;
```

Comment:
I think this code super confusing. In the same function, we use
sdev->ipc_irq
sdev->hda->irq
pci->irq

it wouldn't hurt to clarify what we use when?

Reply:
Actually, in whatever mode, sdev->ipc_irq and sdev->hda->irq are used. ipc_irq is for ipc interrupt and hda->irq is for dsp stream interrupt.

In IO-APIC mode, sdev->ipc_irq and sdev->hda->irq are the same as pci->irq based on our platforms. So you can see pci->irq is assigned to sdev->ipc_irq and sdev->hda->irq.

In MSI mode, we get the irq number by pci_irq_vector(pci, 0).